src/Database/LSMTree/Internal/IndexOrdinaryAcc.hs

```diff
@@ -0,0 +1,100 @@
+
+{-|
+    Incremental construction functionality for the general-purpose fence pointer
+    index.
+-}
+module Database.LSMTree.Internal.IndexOrdinaryAcc
+(
+    IndexOrdinaryAcc,
+    new,
+    append,
+    end
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.List (genericReplicate)
+import           Data.Primitive.ByteArray (byteArrayFromListN)
+import           Data.STRef.Strict (STRef, modifySTRef, newSTRef, readSTRef)
+import           Data.Vector (fromList)
+import qualified Data.Vector.Primitive as Primitive (Vector, length)
+import           Data.Word (Word16, Word8)
+import           Database.LSMTree.Internal.Chunk (Baler, Chunk, createBaler,
+                     feedBaler, readBalerRemnant)
+import           Database.LSMTree.Internal.IndexCompactAcc
+                     (Append (AppendMultiPage, AppendSinglePage))
+import           Database.LSMTree.Internal.IndexOrdinary
+                     (IndexOrdinary (IndexOrdinary))
+import           Database.LSMTree.Internal.Serialise
+                     (SerialisedKey (SerialisedKey'))
+import           Database.LSMTree.Internal.Vector (mkPrimVector)
+
+{-|
+    A general-purpose fence pointer index under incremental construction.
+
+    A value @IndexOrdinaryAcc lastKeysRevRef baler@ denotes a partially
+    constructed index with the following properties:
+
+      * The keys pointed to by @lastKeysRevRef@ are the keys that the index
+        assigns to pages, but in reverse order.
+      * The @baler@ object is used by the index for incremental output of the
+        serialised key list.
+-}
+data IndexOrdinaryAcc s = IndexOrdinaryAcc !(STRef s [SerialisedKey]) !(Baler s)
+
+-- | Creates a new, initially empty, index.
+new :: Int                       -- ^ Maximum chunk size in bytes
+    -> ST s (IndexOrdinaryAcc s) -- ^ Construction of the index
+new maxChunkSize = IndexOrdinaryAcc <$> newSTRef [] <*> createBaler maxChunkSize
+
+{-
+    Appends a single key to the key list of an index and outputs newly available
+    chunks of the serialised key list.
+-}
+appendKey :: SerialisedKey -> IndexOrdinaryAcc s -> ST s [Chunk]
+appendKey lastKey@(SerialisedKey' lastKeyBytes)
+          (IndexOrdinaryAcc lastKeysRevRef baler)
+    = do
+          modifySTRef lastKeysRevRef (lastKey :)
+          chunksIntoLastKeySize <- feedBaler lastKeySizeBytes baler
+          chunksIntoLastKey <- feedBaler lastKeyBytes baler
+          return (chunksIntoLastKeySize ++ chunksIntoLastKey)
+    where
+
+    lastKeySize :: Int
+    lastKeySize = Primitive.length lastKeyBytes
+
+    lastKeySizeAsWord16 :: Word16
+    lastKeySizeAsWord16
+        | lastKeySize <= fromIntegral (maxBound :: Word16)
+            = fromIntegral lastKeySize
+        | otherwise
+            = error "Serialised key too large"
+
+    lastKeySizeBytes :: Primitive.Vector Word8
+    lastKeySizeBytes = mkPrimVector 0 2 $
+                       byteArrayFromListN 1 [lastKeySizeAsWord16]
```

Comment:
It's minor, but this will allocate a list `Cons` and `Nil` for every key. We can probably get around that by creating the byte array ourselves without `byteArrayFromListN`.

Reply:
I made a respective change in 87a9d4d788e980707917730d630425e18ff90fd1. Does this correspond to what you had in mind?